finality-aleph/src/abft/current/performance/scorer.rs

```diff
@@ -0,0 +1,153 @@
+use current_aleph_bft::{NodeCount, NodeMap, Round};
+
+use crate::{abft::current::performance::Batch, aleph_primitives::RawScore, UnverifiedHeader};
+
+/// Scoring ABFT performance based on returned ordered unit batches.
+pub struct Scorer {
+    newest_unit_by: NodeMap<Round>,
+}
+
+impl Scorer {
+    /// Create a new scorer for the provided node count.
+    pub fn new(node_count: NodeCount) -> Self {
+        Scorer {
+            newest_unit_by: NodeMap::with_size(node_count),
+        }
+    }
+
+    /// Add a batch of ordered units and return a score consisting of numbers of rounds a specific
+    /// node is behind.
+    pub fn process_batch<UH: UnverifiedHeader>(&mut self, batch: Batch<UH>) -> RawScore {
+        let max_round = batch.last().expect("batches always contain a head").round;
+        for unit in batch {
+            // Units are always added in order, so any unit created by an honest node
+            // here has a round greater than any that was included earlier.
+            // This is not necessarily true for forkers, but punishing them is fine.
+            self.newest_unit_by.insert(unit.creator, unit.round)
+        }
+        self.newest_unit_by
+            .size()
+            .into_iterator()
+            .map(|node_id| {
+                self.newest_unit_by
+                    .get(node_id)
+                    // All other units have lower round than head, so the saturating_sub is just
+                    // subtraction.
+                    .map(|unit_round| max_round.saturating_sub(*unit_round))
+                    // If we don't have a unit it's the same as having a unit of round equal to -1.
+                    .unwrap_or(max_round + 1)
```

Comment:
Technically, this can overflow, as `max_round` is `u16`  , and `RawScore` is `Vec<u16>`.

Reply:
Changed to `saturating_add`, although if we ever hit rounds so close to `u16::MAX` we are in much more trouble anyway. :P